### ISSUE TITLE
PHP Fatal error:  Could not send result to parent

### ISSUE THREAD

**Issue by SunMar** [amphp/parallel#68]
Hi,

While experimenting a bit I am seeing the following error in my logs:
```
 PHP Fatal error:  Could not send result to parent in /socket/vendor/amphp/parallel/lib/Context/Internal/process-runner.php on line 105
```

I've done some debugging and the root cause is because the child process `process-runner.php` is not stopped gracefully. When the loop ends it's forcefully killed via `Runner->kill()` which causes an Exception and then the error the triggered.

It's reproducible by using the [simple http client](https://github.com/amphp/socket/blob/master/examples/simple-http-client.php) example script. A simple setup to reproduce using Docker:

Run:
```
docker run --rm -it php:7.2 bash -l
```

And then just paste:
```sh
   apt-get update \
&& apt-get install --no-install-recommends -y git unzip \
&& curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin/ --filename=composer \
&& git clone https://github.com/amphp/socket.git \
&& cd socket \
&& composer install --no-dev \
&& echo '[PHP]
error_log=/proc/1/fd/2
' > /usr/local/etc/php/php.ini \
&& (php -S localhost:80 &) \
&& touch index.php \
&& php examples/simple-http-client.php http://localhost/
```

That just creates a clone of this repository, sets the PHP error log to the stdout of PID 1 so the error becomes visible and then launches the example against a local web server started via PHP.

Any idea why the child process isn't stopped properly?

**Comment by kelunik** [comment]
@trowski That's the only thing that comes to my mind, yes.

**Comment by trowski** [comment]
@SunMar I pushed another couple of commits to the `fix-shutdown` branch that registers a shutdown function to gracefully shutdown any running worker pool when the script exits. Please remove any manual call you have to `Pool::shutdown()` and give that branch a try and let me know what issues you find.

**Comment by kelunik** [comment]
@trowski It should probably use wait() there or alternatively Loop::stop(). Also, maybe register another shutdown function in the shutdown function to execute the shutdown as late as possible.

**Comment by SunMar** [comment]
@kelunik  Would `wait()` work in this case? Because I think in this particular scenario the heartbeat prevents the loop from stopping again, wouldn't using `wait()` still keep it waiting for a loop that never ends, since the connection isn't going to be closed anywhere during shutdown? Does it make sense to run the loop again and keep doing things asynchronously during shutdown?

**Comment by trowski** [comment]
@SunMar Pushed another commit using `wait()` instead. The heartbeat is what was keeping the loop running again in the shutdown function. `wait()` will stop the loop once the passed promise is resolved. See what happens now.

**Comment by trowski** [comment]
@SunMar I pushed another commit that moves the shutdown logic to each individual worker instead of the pool. This solved some hanging issues I was seeing with pthreads. Can you see if this still solves the problems you were seeing?

**Comment by SunMar** [comment]
@trowski Still working for me, nothing getting stuck and no errors from the workers :).

**Comment by SunMar** [comment]
@trowski I was looking at the changes and noticed `const SHUTDOWN_TIMEOUT = 1000` in `DefaultPool` wasn't removed, but as far as I can see it's no longer used now that the shutdown is moved to the individual workers.

**Comment by trowski** [comment]
@SunMar Yeah, I noticed that too. I also realized I didn't declare the function in `TaskWorker` as static, so I pushed another commit doing that. Should still work on shutdown, but remove a circular reference that would prevent quick GC.

**Comment by SunMar** [comment]
@trowski It seems I found another case where the workers aren't shutdown properly. When you sent a `SIGTERM` (though could be `SIGINT` too) to the program, then the `register_shutdown_function()`s aren't called :(. So if you exit by doing a `kill` on the main PHP process or simply exit it from `CTRL+C` (which does seem like a pretty normal use case for Amphp since usually you'll start the process and just keep the loop running indefinitely, but there will be a point where the process does need to be stopped for maintenance or whathever), the worker error being unable to write the parent process appears again.

I've implemented by own `Loop::onSignal()` to gracefully shutdown anything running in the loop that I put there. Should I expect Amphp to handle the graceful shutdown of the workers in this scenario too or will I need to add an `onSignal()` myself that handles a shutdown on the worker pool?

**Comment by trowski** [comment]
@kelunik Should we register an `onSignal` handler to catch `SIGINT` and `SIGTERM` to gracefully shutdown each worker?

**Comment by trowski** [comment]
After thinking about it, adding signal handlers is probably not a good idea in library code because we override the default behavior of terminating the process on `SIGINT`/`SIGTERM`. @SunMar I think if you want a graceful shutdown on `CTRL-C`, it's up to you to define a signal handler in your app.

**Comment by SunMar** [comment]
@trowski That makes sense to me. A new library can do its own signal handling, but it's definitely not something that I would automatically assume to be expected behavior.

Then I guess the question is how can I know in the `onSignal()` when it's time to shutdown the workers. Is there some way to know when all promises are resolved, without having to try and track everything myself? And what about (referenced) watchers? Or will `\Amp\Parallel\Worker\pool()->shutdown()` already handle that and is it enough to wrap it in a `wait()` or something?

**Comment by trowski** [comment]
Defining a signal handler for `SIGINT`/`SIGTERM` that ends loop execution (that is, calls `Loop::stop()` or removes all active watchers from the loop) should be enough for the shutdown callback to be invoked. The issue appears to be that the default signal handlers immediately stop executing PHP code, therefore do not allow our shutdown handler to be invoked. You can call `\Amp\Parallel\Worker\pool()->shutdown()` in your signal handler, but I actually do not believe it is necessary now.

Please update to `v1.1.0` that I just tagged and let me know if you have any issues after defining some signal handlers in your app.

**Comment by laurent35240** [comment]
Hi,

I upgrade to version 1.1.0 and I do not have anymore this error.

Thank you for this new version
